# 1.RFM theroy and usage in Marketing Research field
    

RFM analysis is a method used for analysing existing customer and segmenting them into differenct categories so that marketers could target customers within each specific segment more effectively. It is based on simple theory that people are more likely to respond to a new offer if they had purchsed recently and if they are regular customer who had spent heavyly within the retail business. It is mostly used in direct marketing and normally the data should have each transaction in each row and there may be multiple transactions for each costomer..<br>
RFM stands for the three dimensions:<br>
Recency – How recently did the customer purchase? <br>
Frequency – How often did the customer purchase?<br>
Monetary Value – How much did the customer spend?<br>
In order to classify customers according to their RFM behaviour, each customer is assigned a score based on one of the three aspects.For example,if you use a scale of 1-5 and for the customers with the most recent purchase dates you would rank them score of 1 which means they are the best customer in your theory. Consequently, on the same ranking scheme, you could rank customers who purchase most often as score of 1 and who spend the highest monetary values as score of 1.<br>A mock example is created for reference shown below.

![title](img/rfm.png)

# 2. Dataset used for this tutorial and data cleasing

The dataset will be used for this tutorial is from the link below.<br>
http://archive.ics.uci.edu/ml/datasets/online+retail<br>
This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail. Since I only use it for illustration usage, I will only filter out Germany customers for this anslysis.



In [1]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"
import pandas as pd
import numpy as np
import datetime as dt
import sys
sys.version

'3.7.0 (default, Jun 28 2018, 08:04:48) [MSC v.1912 64 bit (AMD64)]'

In [2]:
df=pd.read_excel('Online Retail.xlsx')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
InvoiceNo      541909 non-null object
StockCode      541909 non-null object
Description    540455 non-null object
Quantity       541909 non-null int64
InvoiceDate    541909 non-null datetime64[ns]
UnitPrice      541909 non-null float64
CustomerID     406829 non-null float64
Country        541909 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [4]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

If we divide the data according to countries, we could have a rough idea that UK customers and Germany customers may be the largest two groups for the whole dataset becaue they have relatively large transactions.

In [5]:
df.groupby('Country')['InvoiceNo'].count().sort_values(ascending=False)

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


### I will only use Germany customers for analysis because the data size is relatively small.

In [6]:
Germany=df[df.Country=="Germany"]


In [7]:
Germany.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9495 entries, 1109 to 541805
Data columns (total 8 columns):
InvoiceNo      9495 non-null object
StockCode      9495 non-null object
Description    9495 non-null object
Quantity       9495 non-null int64
InvoiceDate    9495 non-null datetime64[ns]
UnitPrice      9495 non-null float64
CustomerID     9495 non-null float64
Country        9495 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 667.6+ KB


We could see that there are only about 95 unique customers in Germany for the online retail business 

In [8]:
Germany.nunique()

InvoiceNo       603
StockCode      1671
Description    1703
Quantity         68
InvoiceDate     598
UnitPrice       106
CustomerID       95
Country           1
dtype: int64

In [9]:
Germany.describe()

Quantity    UnitPrice    CustomerID
count  9495.000000  9495.000000   9495.000000
mean     12.369458     3.966930  12646.139336
std      17.865719    16.549026    309.138487
min    -288.000000     0.000000  12426.000000
25%       5.000000     1.250000  12480.000000
50%      10.000000     1.950000  12592.000000
75%      12.000000     3.750000  12662.000000
max     600.000000   599.500000  14335.000000

### Since there should not be any data value less than zero in the Quantify variable,I will remove the negative values. It turn out that there is only 1 customer has the negative value.


In [10]:
Germany = Germany[(Germany['Quantity']>0)]
Germany.shape
Germany.nunique()

(9042, 8)

InvoiceNo       457
StockCode      1665
Description    1697
Quantity         47
InvoiceDate     452
UnitPrice       106
CustomerID       94
Country           1
dtype: int64

### create monetary value 'Sales' 

In [11]:
Germany["Sales"]=Germany["Quantity"]*Germany["UnitPrice"]

In [12]:
Germany["UnitPrice"].describe()

count    9042.000000
mean        3.708487
std        13.155279
min         0.000000
25%         1.250000
50%         1.950000
75%         3.750000
max       599.500000
Name: UnitPrice, dtype: float64

### check purchase date and create cutoff date 

In [13]:
Germany["InvoiceDate"].describe()

count                    9042
unique                    452
top       2011-01-07 12:28:00
freq                      149
first     2010-12-01 13:04:00
last      2011-12-09 12:16:00
Name: InvoiceDate, dtype: object

In [14]:
Germany["InvoiceDate"]=pd.to_datetime(Germany["InvoiceDate"])

In [15]:
cutoff_date=dt.datetime(2011,12,10)

In [16]:
df1=Germany.copy()

# 3. RFM data analysis

In [17]:
rfm=df1.groupby('CustomerID').agg({'InvoiceDate':lambda x: (cutoff_date-x.max()).days,
                                 "InvoiceNo": lambda x: len(x),
                                 "Sales": lambda x: x.sum()})

In [18]:
rfm['InvoiceDate']=rfm['InvoiceDate'].astype(int)

In [19]:
rfm.rename(columns={"InvoiceDate":"Recency",
                   "InvoiceNo":"Frequency",
                   "Sales":"MonetaryValue"},inplace=True)
rfm.sort_values('Recency').tail(10)

Recency  Frequency  MonetaryValue
CustomerID                                   
12622.0         232         11         180.10
12585.0         234         87        2040.10
12580.0         246         24         415.70
12623.0         276         18         305.10
12614.0         277         20         315.60
12561.0         302         16         238.85
12559.0         310         21         561.71
12665.0         324          3          63.00
12501.0         336        149        2169.39
12738.0         372         11         155.35

The customer with CustomerID=12665 whose last purchase from the online business was nearly one year ago and for the past year the customer only purchased three times with total monetary value worth $63. Comparing with other frequent customers, this customer is definitely far from being our profitable and loyal customer. Let's check the info in detail for curiosity reason.

In [20]:
low_value_customer=df1[df1["CustomerID"]==12665.0]
low_value_customer

InvoiceNo StockCode                     Description  Quantity  \
62995    541517     21232  STRAWBERRY CERAMIC TRINKET BOX        24   
62996    541517      POST                         POSTAGE         1   
62997    541517     21231  SWEETHEART CERAMIC TRINKET BOX        12   

              InvoiceDate  UnitPrice  CustomerID  Country  Sales  
62995 2011-01-19 08:34:00       1.25     12665.0  Germany   30.0  
62996 2011-01-19 08:34:00      18.00     12665.0  Germany   18.0  
62997 2011-01-19 08:34:00       1.25     12665.0  Germany   15.0

## calculate RFM Scores using percentiles

### Since we need to rank each RFM metirc to calculate respective RFM score, the most convinient way to do so is using percentiles. We will split each metric into 5 segments.

In [21]:
quantile=rfm.quantile(q=[.2,.4,.6,.8])
quantile=quantile.to_dict()
quantile

{'Recency': {0.2: 14.600000000000001,
  0.4: 23.400000000000006,
  0.6: 58.8,
  0.8: 129.4},
 'Frequency': {0.2: 20.0,
  0.4: 36.400000000000006,
  0.6: 68.8,
  0.8: 155.20000000000002},
 'MonetaryValue': {0.2: 428.84000000000003,
  0.4: 748.652,
  0.6: 1683.8000000000002,
  0.8: 3885.2880000000005}}

Since we our definition of best customer is customer  with highest monetary value and highest frequency but lowest recency. We could rank Recency differently with the other two variables.

In [22]:
def R_score(x,percentile,d):
    if x<=d[percentile][.2]:
        return str(1)
    elif x<=d[percentile][.4]:
        return str(2)
    elif x<=d[percentile][.6]:
        return str(3)
    elif x<=d[percentile][.8]:
        return str(4)
    else:
        return str(5)
def FM_score(x,percentile,d):
    if x<=d[percentile][.2]:
        return str(5)
    elif x<=d[percentile][.4]:
        return str(4)
    elif x<=d[percentile][.6]:
        return str(3)
    elif x<=d[percentile][.8]:
        return str(2)
    else:
        return str(1)
    

In [23]:
rfm["R_score"]=rfm["Recency"].apply(R_score,args=('Recency',quantile,))
rfm["F_score"]=rfm["Frequency"].apply(FM_score,args=('Frequency',quantile,))
rfm["M_score"]=rfm["MonetaryValue"].apply(FM_score,args=('MonetaryValue',quantile,))
rfm["RFM_score"]=(rfm["R_score"]+rfm["F_score"]+rfm["M_score"]).astype(int)
rfm=rfm.reset_index()

In [24]:
rfm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 94 entries, 0 to 93
Data columns (total 8 columns):
CustomerID       94 non-null float64
Recency          94 non-null int32
Frequency        94 non-null int64
MonetaryValue    94 non-null float64
R_score          94 non-null object
F_score          94 non-null object
M_score          94 non-null object
RFM_score        94 non-null int32
dtypes: float64(2), int32(2), int64(1), object(3)
memory usage: 5.2+ KB


In [25]:
rfm.sort_values('RFM_score').head(10)

CustomerID  Recency  Frequency  MonetaryValue R_score F_score M_score  \
52     12621.0        1        467       13689.67       1       1       1   
3      12471.0        2        460       19824.05       1       1       1   
83     12720.0        2        354        5781.73       1       1       1   
57     12627.0       10        230        4478.53       1       1       1   
37     12569.0        2        188        4124.69       1       1       1   
71     12668.0       11        161        3807.15       1       1       2   
69     12662.0        0        230        3849.78       1       1       2   
26     12524.0        9        144        4485.72       1       2       1   
29     12528.0        9         92        1933.48       1       2       2   
21     12518.0        0        124        2056.89       1       2       2   

    RFM_score  
52        111  
3         111  
83        111  
57        111  
37        111  
71        112  
69        112  
26        121  
29        122  
21        122

## we could see that only 5 customers out of 94 customers are labeled as '111' which stands for our most valuable and loyal customer.

In [26]:
rfm.sort_values('RFM_score').tail()

CustomerID  Recency  Frequency  MonetaryValue R_score F_score M_score  \
53     12622.0      232         11         180.10       5       5       5   
54     12623.0      276         18         305.10       5       5       5   
35     12561.0      302         16         238.85       5       5       5   
84     12738.0      372         11         155.35       5       5       5   
70     12665.0      324          3          63.00       5       5       5   

    RFM_score  
53        555  
54        555  
35        555  
84        555  
70        555

# 4. Summary
By using RFM analysis, we could divide our retail customers into different segments and by doing so we could promote our products more effectively and improve specific KPIs accordingly.